src/Shared/CopyOnWriteDictionary.cs

```diff
@@ -318,16 +272,11 @@ public bool ContainsKey(K key)
         /// </summary>
         public bool Remove(K key)
         {
-            // Avoid a clone if it's not present
-            if (ReadOperation.HasNoClones || ReadOperation.ContainsKey(key))
-            {
-                if (!IsDummy)
-                {
-                    return WriteOperation.Remove(key);
-                }
-            }
+            ImmutableDictionary<K, V> initial = backing;
 
-            return false;
+            backing = backing.Remove(key);
+
+            return initial != backing; // whether the removal occured
```

Comment:
Should we be cloning?

Reply:
I was thinking the same thing, but calling `Remove` actually does the cloning and returns a new `ImmutableDictionary`